Large NDPI file fix

ntrahearn opened the pull request:
Addresses issue https://github.com/openslide/openslide/issues/174 (and possibly others), regarding the reading of large NDPI files. This issue stems from the fact that classical TIFF (and by extension, NDPI) images only support up to 32 bit values for tagged metadata. However, due to the nature of whole slide image data, it isn't uncommon for the size of an NDPI to exceed the 32 bit range. This means that key metadata, such as the byte positions of image layers or their size in bytes, may be too large to store in a traditional TIFF IFD entry.

Currently OpenSlide relies on heuristics to determine the high bits of 64 bit addresses, which fail in some cases. However, this is unnecessary, as NDPI actually stores the high bits of the offset/value of each tag in 4 byte blocks immediately after the end of the IFD.

This fix modifies openslide-decode-tifflike.c to append these extra 4 bytes to each IFD entry's value/offset and, if necessary, modifies its type to LONG8.

This fix also modifies openslide-vendor-hamamatsu.c to construct correct restart marker addresses. Currently only the values in TIFF tag 65426 are used, which are only the lower 32 bits of each address. High bits are stored in TIFF tag 65432, so these are now appended before mcu_starts are calculated

ntrahearn committed:
hamamatsu: add support for NDPI images > 4 GB

Addresses issue #174 (and possibly others), regarding the reading of
large NDPI files.  This issue stems from the fact that classical TIFF
(and by extension, NDPI) images only support up to 32 bit values for
tagged metadata.  However, due to the nature of whole slide image data,
it isn't uncommon for the size of an NDPI to exceed the 32 bit range.
This means that key metadata, such as the byte positions of image layers
or their size in bytes, may be too large to store in a traditional TIFF
IFD entry.

Currently OpenSlide relies on heuristics to determine the high bits of 64
bit addresses, which fail in some cases.  However, this is unnecessary,
as NDPI actually stores the high bits of the offset/value of each tag in
4 byte blocks immediately after the end of the IFD.

This fix modifies openslide-decode-tifflike.c to append these extra 4
bytes to each IFD entry's value/offset and, if necessary, modifies its
type to LONG8.

This fix also modifies openslide-vendor-hamamatsu.c to construct correct
restart marker addresses.  Currently only the values in TIFF tag 65426
are used, which are only the lower 32 bits of each address.  High bits
are stored in TIFF tag 65432, so these are now appended before mcu_starts
are calculated.

Signed-off-by: Nick Trahearn <n.a.trahearn@gmail.com>
Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
hamamatsu: shorten names of MCU starts constants/variables

Also fix debug print that double-accounted for the MCU starts high half.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
hamamatsu: consolidate unreliable MCU start cases

Drop duplicate code between the low+high and low+no-high cases.

Also logical-or the halves of the offset rather than adding them.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
tifflike: read NDPI value extensions all at once

Avoid seeking twice per directory entry.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
tifflike: minor cleanups

Stop calculating whether value[] should be 4 or 8 bytes: both values are
small and we no longer use sizeof(value) as the read length.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
test/cases: add tests for NDPI file with 4+ GiB bottom level

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert commented:
There's a substantial backlog of PRs I need to go through; I'll get to it but not yet.  I'm working on improving the testing infrastructure first.

joss94 commented:
Hi ! 

Thank you so much for the amazing work on OpenSlide !
Any update on when this PR would be merged ? I guess it would be  really helpful to many people, it sure would help me a lot !

Thanks again

cc-jj commented:
@bgilbert thank you for your work on Openslide. Can you provide an ETA for when this fix will be merged?

carlossantosgarcia commented:
Hello !

Thanks for your work, any updates on this?

bgilbert committed:
tifflike: perform NDPI value extension only on specific field types

We can extend LONG to LONG8 and SLONG to SLONG8, but it isn't clear what's
meant if we see a value extension for other types.  For types with smaller
value sizes, if we overrode the size to 8 when a field's count was greater
than 1, we'd overrun the value[] stack buffer.  Fail on nonempty value
extensions for other types so people will report them to us.

It turns out NDPI sometimes stores short ASCII fields out-of-line (tag
65459 in Hamamatsu-1.ndpi).  Without clarity on whether it always does so,
assume an ASCII field is out-of-line if we see a value extension for it,
and not otherwise for now.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
misc/tiff-delete-tag: port to Python 3

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
test/cases: fix cases broken by NDPI value extension support

Now that we use RAII it's no longer so important that we test every error
path.  -tiff-no-next-directory-offset and -tiff-bad-directory-entry now
both fail reading value extensions; update the error message for one,
since the test already exists, and drop the other.

Test blobs are supposed to be immutable, so rename the refreshed tests.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
tests: update frozen archive

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
hamamatsu: check NDPI_MCU_STARTS_HIGH value count and successful read

Don't use unreliable MCU starts if the high half value count is wrong or
if the high half doesn't read successfully.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert commented:
Thank you again for tracking this down and submitting a fix!  (And for your extreme patience waiting for it to merge.)
